source/common/router/scoped_rds.cc

```diff
@@ -100,71 +108,182 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+        auto* thread_local_scoped_config =
+            const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+        thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
+        return config;
+      });
+      any_applied = true;
+      ENVOY_LOG(debug, "srds: add/update scoped_route '{}'", scoped_route_info->scopeName());
+    } catch (const EnvoyException& e) {
+      exception_msgs.emplace_back(fmt::format("{}", e.what()));
     }
   }
-  for (const auto& scoped_route : scoped_routes) {
-    MessageUtil::validate(scoped_route);
+  if (overriding_init_manager != nullptr) {
+    // For new RDS subscriptions created after listener warming up, we don't wait for them to warm
+    // up.
+    Init::WatcherImpl noop_watcher(
+        // Note: we just throw it away.
+        fmt::format("SRDS ConfigUpdate watcher {}:{}", name_, version_info),
+        []() { /*Do nothing.*/ });
+    overriding_init_manager->initialize(noop_watcher);
+    // New RDS subscriptions should be created, now lift the floodgate.
+    factory_context_.clusterManager().adsMux().resume(
+        Envoy::Config::TypeUrl::get().RouteConfiguration);
   }
 
-  // TODO(AndresGuedez): refactor such that it can be shared with other delta APIs (e.g., CDS).
-  std::vector<std::string> exception_msgs;
-  // We need to keep track of which scoped routes we might need to remove.
-  ScopedConfigManager::ScopedRouteMap scoped_routes_to_remove =
-      scoped_config_manager_.scopedRouteMap();
-  for (auto& scoped_route : scoped_routes) {
-    const std::string& scoped_route_name = scoped_route.name();
-    scoped_routes_to_remove.erase(scoped_route_name);
-    ScopedRouteInfoConstSharedPtr scoped_route_info =
-        scoped_config_manager_.addOrUpdateRoutingScope(scoped_route, version_info);
-    ENVOY_LOG(debug, "srds: add/update scoped_route '{}'", scoped_route_name);
-    applyConfigUpdate([scoped_route_info](const ConfigProvider::ConfigConstSharedPtr& config)
-                          -> ConfigProvider::ConfigConstSharedPtr {
-      auto* thread_local_scoped_config =
-          const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
-      thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
-      return config;
-    });
+  for (const auto& scope_name : removed_resources) {
+    auto iter = scoped_route_map_.find(scope_name);
+    if (iter != scoped_route_map_.end()) {
+      to_be_deleted.push_back(iter->second);
+      scope_name_by_hash_.erase(iter->second->scopeKey().hash());
+      scoped_route_map_.erase(iter);
+      applyConfigUpdate([scope_name](ConfigProvider::ConfigConstSharedPtr config)
+                            -> ConfigProvider::ConfigConstSharedPtr {
+        auto* thread_local_scoped_config =
+            const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+        thread_local_scoped_config->removeRoutingScope(scope_name);
```

Comment:
I wonder if we should be paused around all this code. Won't removing resources cause RDS providers to potentially disappear and changes occur to subscriptions?

Reply:
Done. 
Use a function wide Cleanup object to unbar the new RDS subscriptions.